test/test_common/logging.h

```diff
@@ -61,30 +61,103 @@ class LogRecordingSink : public Logger::SinkDelegate {
+                      "\n Does NOT contain:\n");                                                   \
+      for (const auto& expectation : failed_expectations) {                                        \
+        absl::StrAppend(&failed_message, "  '", expectation.first, "', '", expectation.second,     \
+                        "'\n");                                                                    \
+      }                                                                                            \
+      FAIL() << failed_message;                                                                    \
+    }                                                                                              \
+  } while (false)
+
+// Validates that when stmt is executed, log message containing substr and loglevel will NOT be
+// emitted. Failure message e.g.,
+//
+// Logs:
+// [2018-04-12 05:51:00.245][7290192][warning][upstream] grpc_mux_impl.cc:160] Received gRPC
+// [2018-04-12 05:51:00.246][7290192][trace][upstream] grpc_mux_impl.cc:63] Called bar
+//  Should do NOT contain:
+//   'warning', 'Received gRPC’
+#define EXPECT_LOG_NOT_CONTAINS(loglevel, substr, stmt)                                            \
   do {                                                                                             \
     LogLevelSetter save_levels(spdlog::level::trace);                                              \
     LogRecordingSink log_recorder(Logger::Registry::getSink());                                    \
     stmt;                                                                                          \
-    ASSERT_EQ(1, log_recorder.messages().size());                                                  \
-    std::string recorded_log = log_recorder.messages()[0];                                         \
-    std::vector<absl::string_view> pieces = absl::StrSplit(recorded_log, "][");                    \
-    /* Parse "[2018-04-02 19:06:08.629][15][warn][admin] source/file.cc:691] message ..." */       \
-    ASSERT_LE(3, pieces.size());                                                                   \
-    EXPECT_EQ(loglevel, std::string(pieces[2])); /* error message is legible if cast to string */  \
-    EXPECT_TRUE(absl::string_view(recorded_log).find(substr) != absl::string_view::npos)           \
-        << "\n Actual Log:         " << recorded_log << "\n Expected Substring: " << substr;       \
+    for (const auto& message : log_recorder.messages()) {                                          \
+      if ((message.find(substr) != std::string::npos) &&                                           \
+          (message.find(loglevel) != std::string::npos)) {                                         \
+        FAIL() << "\nLogs:\n"                                                                      \
+               << absl::StrJoin(log_recorder.messages(), "") << "\n Should do NOT contain:\n '"    \
```

Comment:
I definitely think you need to join with something here.  I think `"\n  "` like the prior state's line 87.  But the only way really is to play with it by injecting an error and seeing what looks most readable.

Reply:
per previous comment, joining with "   " would be a little better then.